cloudinit/cmd/main.py

```diff
@@ -1122,8 +1138,53 @@ def main(sysv_args=None):
 
             status_parser(parser_status)
             parser_status.set_defaults(action=("status", handle_status_args))
+    else:
+        parser.error("a subcommand is required")
 
     args = parser.parse_args(args=sysv_args)
+    if not args.single_process:
+        return sub_main(args)
+    LOG.info("Running cloud-init in single process mode.")
+
+    try:
+        # this _must_ be called before sd_notify is called otherwise netcat may
+        # attempt to send "start" before a socket exists
+        sync = socket.SocketSync("local", "network", "config", "final")
+
+        # notify systemd that this stage has completed
+        socket.sd_notify("READY=1")
+
+        # wait for cloud-init-local.service to start
+        with sync("local"):
+            # local stage
+            sub_main(parser.parse_args(args=["init", "--local"]))
+
+        # wait for cloud-init.service to start
+        with sync("network"):
+            # init stage
+            sub_main(parser.parse_args(args=["init"]))
+
+        # wait for cloud-config.service to start
+        with sync("config"):
+            # config stage
+            sub_main(parser.parse_args(args=["modules", "--mode=config"]))
+
+        with sync("final"):
+            # final stage
+            return_code = sub_main(
+                parser.parse_args(args=["modules", "--mode=final"])
+            )
+    except Exception as e:
+        LOG.fatal("Fatal exception: %s", e, exc_info=True)
+        status = traceback.format_exc().replace("\n", " ")
+        socket.sd_notify(f"STATUS={status}")
+        return_code = 1
+    socket.sd_notify("STATUS=Completed")
+    socket.sd_notify("STOPPING=1")
+    return return_code
```

Comment:
Great feedback @aciba90. I'll think about this and propose some changes to address these points.

Reply:
> Summarizing, it feels like
>
> 1. The try-expect should be in a per-stage level
> 2. Non-zero per-stage exit codes should be notified to the init-system
> 3. A more representative exit code should be gathered for cloud-init-single.service

Accomplishing this with only stock `nc.openbsd` isn't possible unfortunately.

To address these concerns, I just pushed this commit: [Improve intra-stage error handling](https://github.com/canonical/cloud-init/pull/5489/commits/1c01d4b4344b6c49644ff8ba49aa62201b02487d). This makes exit codes and error messages possible by allowing cloud-init's Python code to send a short hard-coded shell script on the unix socket which gets piped to a waiting `sh`. This script runs the builtin command `echo` to print the message, and then `exit` to return the appropriate return code.  

Aside: This adds more complexity to the protocol. That said, I think that is faster and simpler to send a two-command shell script to stdin of a waiting shell than it is to parse structured output from the socket in Posix shell. If the current state is too complex, then I think that I would rather treat this nc/shell one-liner as the prototype for a tiny binary in a more fully-featured language (such as Rust) in place of this shell pipeline rather than trying to parse structured output using Posix shell. Even if we are happy with this implementation for now, we may want to replace this with a compiled alternative eventually anyway, given the performance-critical nature of this code.

Reply:
Thanks for handling this. Sounds good to me.